File: src/features/typing/data/java/decode_string.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public String decodeString(String s) {
        Deque<Integer> countStack = new ArrayDeque<>();
        Deque<String> stringStack = new ArrayDeque<>();
        String current = "";
        int count = 0;
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch >= '0' && ch <= '9') {
                count = count * 10 + (ch - '0');
            } else if (ch == '[') {
                countStack.push(count);
                stringStack.push(current);
                current = "";
                count = 0;
            } else if (ch == ']') {
                int repeat = countStack.pop();
                String prev = stringStack.pop();
                for (int j = 0; j < repeat; j++) {
                    prev += current;
                }
                current = prev;
            } else {
                current += ch;
            }
        }
        return current;
    }
}

File: src/features/typing/data/ts/decode_string.ts
function decodeString(s: string): string {
    const countStack: number[] = [];
    const stringStack: string[] = [];
    let current = "";
    let count = 0;
    for (const ch of s) {
        if (ch >= '0' && ch <= '9') {
            count = count * 10 + Number(ch);
        } else if (ch === '[') {
            countStack.push(count);
            stringStack.push(current);
            current = "";
            count = 0;
        } else if (ch === ']') {
            const repeat = countStack.pop()!;
            const prev = stringStack.pop()!;
            let combined = prev;
            for (let j = 0; j < repeat; j++) {
                combined += current;
            }
            current = combined;
        } else {
            current += ch;
        }
    }
    return current;
}
